src/Events/Category_Colors/Migration/Pre_Processor.php

```diff
@@ -0,0 +1,218 @@
+	 *
+	 * @since TBD
+	 * @var array<string, mixed>
+	 */
+	protected array $processed_settings = [];
+
+	/**
+	 * Processes category colors and settings for migration.
+	 * Fires an action before and after processing.
+	 * If the process completes successfully, the end hook passes `true`.
+	 * If processing is skipped due to empty settings, the end hook passes `false`.
+	 *
+	 * @since TBD
+	 * @return bool
+	 */
+	public function process(): bool {
+		$start_time = microtime( true );
+		$this->update_migration_status( Status::$in_progress );
+
+		/**
+		 * Fires before the preprocessor starts processing category color data.
+		 * Allows logging or hooking into the process before any changes are made.
+		 *
+		 * @since TBD
+		 */
+		do_action( 'tec_events_category_colors_migration_preprocessor_start' );
+
+		// Load the original settings.
+		$this->processed_settings = $this->get_original_settings();
+
+		if ( empty( $this->processed_settings ) ) {
+			$this->update_migration_data( Config::$expected_structure );
+			$this->update_migration_status( Status::$preprocess_skipped );
+
+			/**
+			 * Fires after the preprocessor completes.
+			 *
+			 * @since TBD
+			 *
+			 * @param array<string, mixed> $migration_data The processed migration data.
+			 * @param bool                 $success        Whether processing was successful.
+			 */
+			do_action( 'tec_events_category_colors_migration_preprocessor_end', Config::$expected_structure, false );
+			$this->log_elapsed_time( 'Preprocessing', $start_time );
+			return false;
+		}
+
+		// Populate migration data.
+		$migration_data = [
+			'categories'    => $this->get_category_values(),
+			'settings'      => $this->get_settings_values(),
+			'ignored_terms' => $this->process_ignored_terms(),
+		];
+
+		// Store processed data in the database.
+		$this->update_migration_data( $migration_data );
+
+		$this->update_migration_status( Status::$preprocess_completed );
+
+		/**
+		 * Fires after the preprocessor completes.
+		 *
+		 * @since TBD
+		 *
+		 * @param array<string, mixed> $migration_data The processed migration data.
+		 * @param bool                 $success        Whether processing was successful.
+		 */
+		do_action( 'tec_events_category_colors_migration_preprocessor_end', $migration_data, true );
+
+		$this->log_message( 'info', 'Preprocessing complete. Migration data prepared.', $migration_data, 'Pre_Processor' );
+
+		$this->log_elapsed_time( 'Preprocessing', $start_time );
+
+		return true;
+	}
+
+	/**
+	 * Extracts category-related values from processed settings and removes them.
+	 *
+	 * @since TBD
+	 * @return array<int, array<string, mixed>> Processed category data structured by category ID.
+	 */
+	protected function get_category_values(): array {
+		$categories        = [];
+		$filtered_settings = $this->processed_settings;
+
+		foreach ( $this->processed_settings['terms'] ?? [] as $term_id => [$slug, $name] ) {
```

Comment:
there's a lot of nesting going on here...maybe array_map and a separate function?